zebra-consensus/src/transaction.rs

```diff
@@ -12,11 +12,283 @@
 //! This is an internal module. Use `verify::BlockVerifier` for blocks and their
 //! transactions, or `mempool::MempoolTransactionVerifier` for mempool transactions.
 
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use displaydoc::Display;
+use futures::{stream::FuturesUnordered, FutureExt};
+use thiserror::Error;
+use tower::Service;
+use tower::ServiceExt;
+
+use zebra_chain::{
+    parameters::NetworkUpgrade,
+    primitives::{ed25519, redjubjub},
+    transaction::{self, HashType, Transaction},
+};
+
+use zebra_state as zs;
+
+use crate::{primitives::groth16, script, BoxError};
+
+mod check;
+
 /// Internal transaction verification service.
 ///
-/// After verification, the transaction future completes. State changes are handled by
-/// `BlockVerifier` or `MempoolTransactionVerifier`.
+/// After verification, the transaction future completes. State changes are
+/// handled by `BlockVerifier` or `MempoolTransactionVerifier`.
+#[derive(Debug, Clone)]
+pub struct Verifier<ZS> {
+    script_verifier: script::Verifier<ZS>,
+    // spend_verifier: groth16::Verifier,
+    // output_verifier: groth16::Verifier,
+    // joinsplit_verifier: groth16::Verifier,
+}
+
+impl<ZS> Verifier<ZS>
+where
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
+    ZS::Future: Send + 'static,
+{
+    pub(crate) fn new(
+        script_verifier: script::Verifier<ZS>,
+        // spend_verifier: groth16::Verifier,
+        // output_verifier: groth16::Verifier,
+        // joinsplit_verifier: groth16::Verifier,
+    ) -> Self {
+        Self {
+            script_verifier,
+            // spend_verifier,
+            // output_verifier,
+            // joinsplit_verifier,
+        }
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Display, Error)]
+pub enum VerifyTransactionError {
+    /// {0} version is older than the minimum transaction version: V4
+    WrongVersion(String),
+    /// transaction transfers no money
+    NoTransfer,
```

Comment:
This is a misleading name, as it suggests that the consensus rules verify that zero-value coins and notes cannot be spent. The actual rules that lead to this error are only checking for the presence of inputs / spends.